### PR TITLE
[WIP] Torch operator and NDArray function

### PR DESCRIPTION
**Pull request by piiswrong**
this is currently working and can be tested with 
```Python
import ctypes
lualib = ctypes.CDLL("libluajit.so", mode=ctypes.RTLD_GLOBAL)
import mxnet as mx
import numpy as np
from mxnet.base import FunctionHandle, mx_uint, check_call, _LIB, py_str, c_array, NDArrayHandle, mx_float
# data = mx.symbol.Variable('data')
# f32data = mx.symbol.Cast(data=data, dtype='float32')
# fc1  = mx.symbol.FullyConnected(data = f32data, name='fc1', num_hidden=128)
# mlp  = mx.symbol.SoftmaxOutput(data = fc1, name = 'softmax')

# print mlp.infer_shape(data=(128,77))
# print mlp.infer_type(data=np.float16)

# exe = mlp.simple_bind(mx.gpu(0), 'write', {'data': np.float32}, data=(128,77))
# exe.forward(is_train=True)
# exe.backward()

# print exe.grad_arrays
# for arr in exe.grad_arrays:
#     if arr:
#         arr.wait_to_read()

plist = ctypes.POINTER(FunctionHandle)()
size = ctypes.c_uint()
check_call(_LIB.MXListFunctions(ctypes.byref(size),
                                ctypes.byref(plist)))
for i in range(size.value):
    hdl = FunctionHandle(plist[i])
    n_used_vars = mx_uint()
    n_scalars = mx_uint()
    n_mutate_vars = mx_uint()
    type_mask = ctypes.c_int()
    check_call(_LIB.MXFuncDescribe(
        hdl,
        ctypes.byref(n_used_vars),
        ctypes.byref(n_scalars),
        ctypes.byref(n_mutate_vars),
        ctypes.byref(type_mask)))
    n_mutate_vars = n_mutate_vars.value
    n_used_vars = n_used_vars.value
    n_scalars = n_scalars.value
    type_mask = type_mask.value

    name = ctypes.c_char_p()
    desc = ctypes.c_char_p()
    num_args = mx_uint()
    arg_names = ctypes.POINTER(ctypes.c_char_p)()
    arg_types = ctypes.POINTER(ctypes.c_char_p)()
    arg_descs = ctypes.POINTER(ctypes.c_char_p)()

    check_call(_LIB.MXFuncGetInfo(
        hdl, ctypes.byref(name), ctypes.byref(desc),
        ctypes.byref(num_args),
        ctypes.byref(arg_names),
        ctypes.byref(arg_types),
        ctypes.byref(arg_descs)))
    func_name = py_str(name.value)

    if func_name == '_torch_add':
        break
print func_name

x = mx.nd.zeros((10,10), ctx=mx.gpu(2))
x[:] = -3
x.wait_to_read()
print x.asnumpy()
check_call(_LIB.MXFuncInvoke( \
            hdl, \
            c_array(NDArrayHandle, [arg.handle for arg in [x]]), \
            c_array(mx_float, []), \
            c_array(NDArrayHandle, [v.handle for v in [x]]),
            ctypes.c_int(2),
            c_array(ctypes.c_char_p, ['format', 'args']),
            c_array(ctypes.c_char_p, ['nnf', ',,2']),
            ))

print x.asnumpy()
print x.context
```
Still need to work on python side interface.

### CODE REVIEW

**Comment by piiswrong** [comment]
@tqchen could you review this?

File: include/mxnet/ndarray.h [hunk 1]
```diff
@@ -554,6 +559,27 @@ struct NDArrayFunctionReg
...
+   * \brief set the function body to a unary NDArray function
+   *  this will also auto set the parameters correctly
+   * \param fgeneric function body to set
+   * \return ref to the registered entry, used to set properties
+   */
+  inline NDArrayFunctionReg &set_function(
+    void (*fgeneric)(NDArray **used_vars,
+                     real_t *s,
+                     NDArray **mutate_vars,
+                     const std::map<std::string, std::string>& param)) {
```
Comment (tqchen) [inline include/mxnet/ndarray.h]: try pass char** char**, to reduce the cost of constructing the map, note this call can have significant amount of overhead

File: python/mxnet/torch.py [hunk 2]
```diff
@@ -0,0 +1,157 @@
...
+"""Interface for NDArray functions executed by torch backend.
+Install torch and Compile with USE_TORCH=1 to use this module"""
+from __future__ import absolute_import
+
+import ctypes
+import sys
+from .base import _LIB
+from .base import c_array, py_str, ctypes2docstring
+from .base import mx_uint, mx_float, NDArrayHandle, FunctionHandle
+from .base import check_call
```
Comment (tqchen) [inline python/mxnet/torch.py]: Can we make this special case of general functions? It is ideal that we do not need torch.py, but ensemble these into the normal mxnet library.

**Comment by tqchen** [comment]
The general structure LGTM, some major comments:

- Let us separate build rule for plugin, this helps later extensions.
   - Possibly with a separate Makefile
- Let us implement all the features without torch.py. I guess it is doable This allows things to be more easily incorporated into other front-end. e.g Julia and R more easily.

Plugin option:
- Compile libmxnet_torch.so separately(which links to libluajit.so) and when loaded, it presents all the features.

**Comment by tqchen** [comment]
also cc @mavenlin on reviewing and @pluskid @thirdwing on language interfacing issues

File: include/mxnet/ndarray.h (hunk 1 above)
Comment (piiswrong) [inline include/mxnet/ndarray.h]: but the map need to be constructed down stream anyway

File: python/mxnet/torch.py (hunk 2 above)
Comment (piiswrong) [inline python/mxnet/torch.py]: It's going to override ndarray functions with the same name

**Comment by piiswrong** [comment]
@tqchen How should libmxnet_torch.so be installed?

**Comment by tqchen** [comment]
I guess we could simply use ctypes to load in the dll, and things should be registered when they are first loaded, then call the init module again to refresh the exposed symbols, But I have not tested it

File: python/mxnet/torch.py (hunk 2 above)
Comment (tqchen) [inline python/mxnet/torch.py]: Hmm, is it possible to simply not override, but expose them as different names? The idea is to have torch as a complement of what we have here

**Comment by piiswrong** [comment]
Let's merge this first and setup plugin system later. I'll talk with people from facebook and see what they think.

**Pull request merged**